rfd/0067-database-access-aws-redis

```diff
@@ -0,0 +1,260 @@
+RFD](https://github.com/gravitational/teleport/blob/master/rfd/0011-database-access.md),
+mutual TLS is used between database clients and proxy, and between proxy and
+database service. The same applies to Redis access.
+
+However, neither mutual TLS nor IAM authentication is supported by Redis
+managed by AWS. This document mainly focuses on access from database service to
+Redis managed by AWS.
+
+## Details
+
+### In-transit encryption (TLS)
+
+For security purposes, Teleport should ONLY
+support databases with in-transit encryption (TLS) enabled.
+
+[Enabling TLS on ElastiCache
+Redis](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/in-transit-encryption.html#in-transit-encryption-enable)
+can be a complicated process, and all client applications must be updated at
+the same time. Therefore, Teleport should NOT automatically enable TLS for
+existing clusters. Redis servers without TLS are NOT supported, and users are
+recommended to enable TLS separately to enable Teleport database access.
+
+### Auth
+
+Redis can be configured with one of the following authentication methods, and
+only one at a time.
+
+#### Auth - Redis with no auth
+
+Redis can be configured without any auth method.
+
+"default" user is always used in this configuration, and no `auth` command is
+required upon a successful connection.
+
+#### Auth - Redis with Redis Auth
+
+[Redis Auth](https://redis.io/commands/auth/)
+uses a single token/password that must be shared for all client applications.
+
+"default" user is always used in this configuration, and users have to manually
+input `auth` command upon a successful connection. <sup>1<sup>
+
+#### Auth - Redis with ACL
+
+[Redis ACL](https://redis.io/docs/manual/security/acl/) (also known as RBAC for
+ElastiCache Redis) is introduced in Redis 6. Both ElastiCache and MemoryDB
+provide APIs to manage database users and user groups for ACL.
+
+As with all other databases, the database users should be created by "database
+admins" (by users, not Teleport) to assign desired permissions. 
+
+To provide better security than using static passwords, users can tag
+ElastiCache and MemoryDB users with label **`teleport.dev/managed: true`** for
+Teleport to manage the passwords for those database users.
+
+Database service discovers database users by listing users from the ElastiCache
+user group or MemoryDB ACL attached to the Redis database. For database users
+discovered with label **`teleport.dev/managed: true`**, passwords are randomly
+generated, periodically rotated, and securely stored in the secret store (more
+on the secret store in a later section).
+
+In HA mode, if more than one database service tries to rotate the password for
+the same user at the same time, the secret store implementation guarantees only
+the first caller succeeds. And only the successor should call AWS APIs to
+update the database user's passwords.
+
+It has been found that it may take a few minutes for AWS to propagate user
```

Comment:
I think this is a very valid concern. Regardless of how often the password is rotated, there is a chance the user connects right before a rotation and this may happen. When creating the Redis client, instead of giving it username and password, I think we can hook up the `OnConnect` handler and fetch-secret-then-auth on the fly. I will give it a try today with the redis client in cluster mode

Reply:
Makes sense. I remember that there was a reason why I didn't use `OnConnect` but I cannot recall it right now. We can go back to it during the implementation.